lib/segment/src/vector_storage/dynamic_mmap_flags.rs

```diff
@@ -0,0 +1,207 @@
+fn mmap_max_current_size(len: usize) -> usize {
+    let mmap_capacity_bytes = mmap_capacity_bytes(len);
+    mmap_capacity_bytes * 8
+}
+
+impl DynamicMmapFlags {
+    fn mmap_file_path(directory: &Path) -> PathBuf {
+        directory.join(FLAGS_FILE)
+    }
+
+    pub fn open(directory: &Path) -> OperationResult<Self> {
+        create_dir_all(directory)?;
+        let status_mmap = ensure_status_file(directory)?;
+        let len = *read_from_mmap(&status_mmap, 0);
+
+        let status = MmapStatus { len };
+
+        let mut flags = Self {
+            status,
+            _flags_mmap: None,
+            flags: None,
+            status_mmap,
+            directory: directory.to_owned(),
+        };
+
+        flags.reopen_mmap(len)?;
+
+        Ok(flags)
+    }
+
+    pub fn reopen_mmap(&mut self, length: usize) -> OperationResult<()> {
+        self.flags.take(); // Drop the bit slice
+        self._flags_mmap.take(); // Drop the mmap
+        let capacity_bytes = mmap_capacity_bytes(length);
+        let mmap_path = Self::mmap_file_path(&self.directory);
+        create_and_ensure_length(&mmap_path, capacity_bytes)?;
+        let mut flags_mmap = open_write_mmap(&mmap_path).describe("Open mmap flags for writing")?;
+        #[cfg(unix)]
+        if let Err(err) = flags_mmap.advise(memmap2::Advice::WillNeed) {
+            log::error!("Failed to advise MADV_WILLNEED for deleted flags: {}", err,);
+        }
+        let flags = unsafe { mmap_to_bitslice(&mut flags_mmap, 0) };
+        self._flags_mmap = Some(flags_mmap);
+        self.flags = Some(flags);
+        Ok(())
+    }
+
+    /// Set the length of the vector to the given value.
+    /// If the vector is grown, the new elements will be set to `false`.
+    /// Errors if the vector is shrunk.
+    pub fn set_len(&mut self, new_len: usize) -> OperationResult<()> {
+        debug_assert!(new_len >= self.status.len);
+        if new_len == self.status.len {
+            return Ok(());
+        }
+
+        let current_capacity = mmap_max_current_size(self.status.len);
+
+        if new_len > current_capacity {
+            self.reopen_mmap(new_len)?;
+        }
+
+        if new_len < self.status.len {
+            return Err(OperationError::service_error(format!(
+                "Cannot shrink the mmap flags from {} to {}",
+                self.status.len, new_len
+            )));
+        }
+
+        self.status.len = new_len;
+        write_to_mmap(&mut self.status_mmap, 0, self.status.len);
+        Ok(())
+    }
+
+    pub fn get<TKey>(&self, key: TKey) -> bool
+    where
+        TKey: num_traits::cast::AsPrimitive<usize>,
+    {
+        let key: usize = key.as_();
+        self.flags.as_ref().map(|flags| flags[key]).unwrap_or(false)
+    }
+
+    /// Set the `true` value of the flag at the given index.
+    /// Ignore the call if the index is out of bounds.
+    ///
+    /// Returns previous value of the flag.
+    pub fn set_true<TKey>(&mut self, key: TKey) -> bool
```

Comment:
`set_true` looks like usage-specific stuff. while reading this file it's unclear why we propose mmap flags container that can set only true values. I propose to rename as `set` and provide `bool` argument

Reply:
:+1: